### ISSUE TITLE
Reference details not available in API

### ISSUE THREAD

**Issue by KevinMitchell** [RepreZen/KaiZen-OpenApi-Parser#75]
In the overview of the parser is says "References are normally traversed automatically by the API, but full details of references and their resolution status are also available."  Are there any examples of how to do this?  For example, from a media type I can get the schema.  But this might be an inlined schema or a reference to a component schema.  But it's not obvious to me how to tell which of these cases applies in a particular instance, and how to get details of the reference.  I can see there are Reference and JsonOverlay classes.  But if I get a Schema instance from a MediaType instance, does the schema instance "know" it was obtained via a reference, and can somehow provide information about the reference?  Which presumably applies that every reference to a component schema will return a different schema instance?

Any hints would be appreciated
Kevin

**Comment by andylowry** [comment]
@KevinMitchell Apologies, but this is not yet implemented, and the "Current State" in the README fails to mention this.

I'm imagining, at minimum, that support will include the following:

* Add a new property to object field definitions in `types3.yaml`, specifying that a reference object is allowed for the field.
* Add new `xxxIsReference()` boolean method to the generated API for these fields, indicating whether field `xxx` is specified as a reference
* Add new `getXxxReference()` method that returns an object from which reference details can be obtained, including:
  * original reference string
  * whether the reference was successfully resolved
  * if not resolved, error messages as to why resolution failed
  * if resolved, fully resolved URI

Anything else you think would be needed?

**Comment by andylowry** [comment]
@KevinMitchell Another esoteric issue with this is that the OpenAPI 3.0 spec defines a `Link` property called `operationRef` which, as far as I can make, out, expects a reference string as its value. Not a reference _object_, but a plain reference _string_. It's a very unfortunate special case of reference handling (one of two - the other being the very odd semantics of `Path` references).

I expect that I'll come up with an API treatment for reference objects first, and put off `Link.operationRef` til the future, unless I can come up with a clever approach.

**Comment by andylowry** [comment]
Maybe something different. I'd have to go back and look more closely. The basic idea of the overlay is to avoid the cost of deserializing the model document into a set of POJOs, and instead create a representation that "overlays" a jackson `JsonNode` structure. I've tried to be careful to make sure that nothing in the 'jsonoverlay' package tree knows anything about OpenAPI so that it can serve as a generic framework for creating parsers for JSON/YAML-based DSLs. 

References, thankfully, are generic JSON technology, hence adding support in the framework is appropriate. Odd cases like `Link.operationRef` don't make sense in the framework, but because of that case I may need to create some hack that crosses the framework boundary in some ugly way. Maybe some sort of hook mechanism that would support restructuring a JSON tree prior to elaborating the overlay structure? (E.g. turning these ref-like properties into ref objects under the covers) would suffice, so the framework would still know nothing about these odd cases? Or maybe just providing reference services generically, and adding manual code to the generated code for `Link` objects would be better. Decisions, decisions... :-)

**Comment by andylowry** [comment]
@JPnCo Mentioning you here so you can follow any progress on this thread. Also please see the discussion above, and chime in if you have any thoughts on the form you'd like the reference API to take. Thanks.

**Comment by KevinMitchell** [comment]
Suppose I define 

```
components:
  schemas:
    MailAddress: …
```

And then elsewhere
```
        mailAddresses:
          type: array
          items:
            $ref: '#/components/schemas/MailAddress'
```
Until we have support for references we can’t “see” where schemas were obtained by following references.  But I thought maybe I could compare the hashcode of the schema in the component with the schema returned as the items schema in the mailAddresses property to detect the references.  But these are different, presumably because you are creating new instances of each component schema when it’s referenced.  I can’t use the title, as that’s not guaranteed to be unique.  So would it make sense to add some kind of schema ID that was guaranteed to be preserved when a schema was referenced and “inlined”?  Or is there some existing property that’s preserved when instantiating a reference?  Obviously I could do a full equality check whenever I get a schema, but that starts to get expensive.

**Comment by KevinMitchell** [comment]
Ah, ignore that last comment.  I see the latest builds of the parser have reference support :-)

**Comment by andylowry** [comment]
@KevinMitchell I expect you're referring to the changes in #85, where there was some attention paid to interfaces. The `isReference` and `getReference` methods are a help, but I'm currently working on some other important improvements, mostly but not entirely to address #64. Among things you should see shortly:

* Reference recursion does not break parser (it will probably still break simple serialization with reference following (`toJson(true)`)
* Original reference strings available from `Reference` objects (currently `getRefString` provides a "canonicalized" reference URI
* Shared overlay instances for multiple references with identical canonical reference strings

This is a little more involved than it might at first seem since I need to switch to a static factory method based instantiation approach for all overlays. I can't, in the middle of a constructor, notice that I'm looking at a reference to an object that's already been overlaid elsewhere and return that previously constructed overlay object! Side-benefit is that this will probably form a more solid basis for the promised builder API, though that's not yet near the top of my priorities.

In the meantime, your need to recognize when two objects really represent the same object can addressed partly by checking whether they're both references and that they have `equals` reference strings. "Partly" because:

* It won't help you in testing that a reference to `#/components/schema/Foo` is the same object as `model.getSchema("Foo")`, since the latter is (probably) not a reference at all. Once I'm done with current work, you'll be able to use `Object.equals` to test this.
* This currently only applies to values of types derived from `ObjectOverlay`, since that's the only case where the "value" type and the "overlay" type are the same. E.g. can't currently check whether the model title string is obtained via a reference, because there's no way to get at the `StringOverlay` object that represents the title. (This will not be a problem for any valid OpenAPI model, since the spec prohibits reference nodes in most places, but the KaiZen parser is lenient in this regard and can handle references anywhere. There'll probably at some point be warnings or errors generated for references not specifically permitted by the OAS specification.)

**Comment by andylowry** [comment]
@KevinMitchell I've just pushed changes that address your need for reference visibility. Per the comments above, this involved a major redesign of some parts of the implementation, due to my mistaken ideas about how references could be exposed in the API.

There's also a new top-level markdown document, `API-Overview`, that explains the API in considerable detail, including treatment of references.

I hope you find the changes helpful.

**Comment by andylowry** [comment]
Addressed in #94
